File: api/convert-yt-mp3.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';
import path from 'path';
import os from 'os';
import fs from 'fs';
import crypto from 'crypto';
import { execFile } from 'child_process';
import { promisify } from 'util';

export const maxDuration = 10;

const execFileAsync = promisify(execFile);

export default async function handler(req: VercelRequest, res: VercelResponse) {
    if (req.method !== 'POST') return res.status(405).end();

    let actualFile = '';
    try {
        const { youtubeUrl } = req.body;
        const urlObj = new URL(youtubeUrl);
        const videoId = urlObj.searchParams.get('v');
        const cleanUrl = videoId ? `https://www.youtube.com/watch?v=${videoId}` : youtubeUrl;

        const tmpDir = os.tmpdir();
        const tmpId = crypto.randomBytes(8).toString('hex');
        const tmpFileTemplate = path.join(tmpDir, `lexis-mp3-${tmpId}.%(ext)s`);
        const ytdlpBin = path.join(process.cwd(), 'node_modules', 'youtube-dl-exec', 'bin', 'yt-dlp');

        let title = `youtube_audio_${tmpId}`;
        try {
            const { stdout } = await execFileAsync(ytdlpBin, [cleanUrl, '--get-title', '--no-warnings', '--no-check-certificates', '--no-playlist'], { timeout: 15000 });
            const t = stdout.trim();
            if (t) title = t.replace(/[^a-zA-Z0-9 _\-]/g, '').trim() || title;
        } catch (_) { }

        await execFileAsync(ytdlpBin, [
            cleanUrl, '--output', tmpFileTemplate,
            '--format', 'bestaudio/best', '--extract-audio',
            '--audio-format', 'mp3', '--audio-quality', '0',
            '--no-warnings', '--no-check-certificates', '--no-playlist'
        ], { timeout: 8000 });

        const tmpFiles = fs.readdirSync(tmpDir).filter(f => f.startsWith(`lexis-mp3-${tmpId}`) && f.endsWith('.mp3'));
        if (!tmpFiles.length) throw new Error('Conversion failed — no output file found.');

        actualFile = path.join(tmpDir, tmpFiles[0]);
        const safeFilename = encodeURIComponent(`${title}.mp3`);

        res.setHeader('Content-Type', 'audio/mpeg');
        res.setHeader('Content-Disposition', `attachment; filename*=UTF-8''${safeFilename}`);
        res.setHeader('Content-Length', String(fs.statSync(actualFile).size));

        const stream = fs.createReadStream(actualFile);
        stream.pipe(res);
        stream.on('end', () => { try { fs.unlinkSync(actualFile); } catch (_) { } });
        stream.on('error', (e) => {
            if (!res.headersSent) res.status(500).json({ error: e.message });
            try { fs.unlinkSync(actualFile); } catch (_) { }
        });
    } catch (err: any) {
        console.error('Convert error:', err.message);
        if (!res.headersSent) res.status(500).json({ error: err.message });
        if (actualFile && fs.existsSync(actualFile)) fs.unlinkSync(actualFile);
    }
}

File: api/convert-yt-mp3.py
import os
import tempfile
import base64
import urllib.parse
from flask import Flask, request, send_file, jsonify
import yt_dlp

app = Flask(__name__)

@app.route('/api/convert-yt-mp3', methods=['POST'])
def handler():
    try:
        data = request.get_json()
        youtube_url = data.get('youtubeUrl')
        
        if not youtube_url:
            return jsonify({"error": "No URL provided"}), 400

        # Create a temporary directory
        tmp_dir = tempfile.mkdtemp()
        file_id = base64.b64encode(os.urandom(6)).decode('utf-8').replace('/', '_').replace('+', '_')
        file_path = os.path.join(tmp_dir, f'lexis_{file_id}')

        ydl_opts = {
            'format': 'bestaudio/best',
            'outtmpl': f'{file_path}.%(ext)s',
            'postprocessors': [{
                'key': 'FFmpegExtractAudio',
                'preferredcodec': 'mp3',
                'preferredquality': '192',
            }],
            'quiet': True,
        }

        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            # Get video title
            info = ydl.extract_info(youtube_url, download=True)
            title = info.get('title', 'audio')
            # Sanitize title
            safe_title = "".join([c for c in title if c.isalnum() or c in (' ', '-', '_')]).strip()

        actual_file = f'{file_path}.mp3'
        
        if not os.path.exists(actual_file):
            return jsonify({"error": "Conversion failed"}), 500

        # Send the file
        response = send_file(
            actual_file,
            mimetype="audio/mpeg",
            as_attachment=True,
            download_name=f"{safe_title}.mp3"
        )
        
        # Add headers for PWA/Vercel
        response.headers["Content-Disposition"] = f"attachment; filename*=UTF-8''{urllib.parse.quote(safe_title)}.mp3"
        return response

    except Exception as e:
        print(f"Error: {str(e)}")
        return jsonify({"error": "Conversion failed. Please try a shorter video."}), 500
